conbench/entities/benchmark_result.py

```diff
@@ -72,26 +84,59 @@ def create(data):
         benchmark_result_data = {}
-        complete_iterations = False
+
+        # Note(JP): if no data was provided, should this stay `True`?
+        # We will see. https://github.com/conbench/conbench/issues/803.
+        any_errored_iteration = True
 
         if has_stats:
-            benchmark_result_data = data["stats"]
+            # Note(JP): this is the result of marshmallow
+            # deserializaiton/validation against the `BenchmarkResultCreate`
+            # schema below. This needs better naming. Most importantly, we
+            # should make use of typing information derived from the schema.
+            benchmark_result_data = data["stats"].copy()
+
+            # Note(JP): I think we may want to emit a Bad Request response when
+            # benchmark_result_data['data'] does not contain any non-null
+            # value. An empty list would then also be rejected right away. `if
+            # benchmark_result_data.get("data")` is probably supposed to catch
+            # the case where an empty list was provided.
 
             # calculate any missing stats if data available
-            if data["stats"].get("data"):
-                dat = [to_float(x) for x in benchmark_result_data["data"]]
+            if benchmark_result_data.get("data"):
+                # Note(JP): looks like `benchmark_result_data["data"]` is a
+                # list of either `None` or `Decimal` objects (currently
+                # enforced via marshmallow schema). Example:
+                #
+                # [Decimal('0.099094'), None, Decimal('0.036381')]
+                #
+                # TODO: let's lock this in with type annotations so that this
+                # is known _here_. Transform `None` to `math.nan` and `Decimal``
+                # to `float`` for subsequent calculations.
+                dat = [
+                    float(x) if x is not None else math.nan
+                    for x in benchmark_result_data["data"]
+                ]
 
-                # defaults
                 q1 = q3 = mean = median = min = max = stdev = iqr = None
 
-                # calculate stats only if the data is complete
-                complete_iterations = all([x is not None for x in dat])
-                if complete_iterations:
+                # calculate stats only if the data is complete. Note(JP): why?
+                # maybe we want to calculate that always if there are at least
+                # N (3?) data points.
+                any_errored_iteration = any([x is math.nan for x in dat])
+
+                if not any_errored_iteration:
+                    # Note(JP): do we really want to populate all of the below
+                    # except for stddev if the sample size is 1? In that case,
+                    # all of max, min, mean, q1 share the same value. I'd say
+                    # it makes more sense to calculate these statistics only if
+                    # three or more samples have been provided.
```

Comment:
related: https://github.com/conbench/conbench/issues/802

Reply:
I love this, absolutely we should do this (happy to have it done now if it makes what you're working on easier, or we can make a ticket to do it later). 

Part of the trouble here ends up being that we use the mean to do many of our subsequent calculations and visualizations: I suspect (like you mention in #802 we probably should "just" set this to calculate these if there are 2 or more (I agree having a nudge towards higher iterations is good, and maybe we can take that stand now, but it would be weird to document ~"you get no stats if you have 2 or less, and therefore we can't do anything with regressions (since we use the mean to calculate that, and it's no calculated). But if it's 1, then we do actually just use that value so it's fine, but you really should be sending 3" Probably best to just have 1 iteration, that's the measure that's used, and then anything more than that we calculate stats like standard

Reply:
Thanks for the feedback! I'd say let's build "calculate these statistics only if three or more samples have been provided" and then adjust consumers where necessary.